common/src/Model/GroupNode.cpp

```diff
@@ -35,59 +35,146 @@
                         auto entity = entityNode->entity();
                         entity.transform(transformation);
-                        return std::make_unique<EntityNode>(std::move(entity));
+                        origNodeAndTransformedContents[i] = {entityNode, std::move(entity)};
                     },
-                    [&](const BrushNode* brushNode) -> VisitResult {
+                    [&](const Model::BrushNode* brushNode) {
                         auto brush = brushNode->brush();
-                        return brush.transform(worldBounds, transformation, true)
-                            .and_then([&]() -> kdl::result<std::unique_ptr<Node>, BrushError> {
-                                return std::make_unique<BrushNode>(std::move(brush));
-                            }).map_errors([](const BrushError&) -> VisitResult { 
-                                return UpdateLinkedGroupsError::TransformFailed;
+                        brush.transform(worldBounds, transformation, true)
+                            .and_then([&]() {
+                                origNodeAndTransformedContents[i] = {brushNode, std::move(brush)};
+                            })
+                            .handle_errors([&](BrushError){
+                                transformFailed = true;
                             });
                     },
-                    [&](const PatchNode* patchNode) -> VisitResult {
+                    [&](const Model::PatchNode* patchNode) {
                         auto patch = patchNode->patch();
                         patch.transform(transformation);
-                        return std::make_unique<PatchNode>(std::move(patch));
+                        origNodeAndTransformedContents[i] = {patchNode, std::move(patch)};
                     }
-                )).and_then([&](std::unique_ptr<Node>&& newChildNode) -> VisitResult {
-                    if (!worldBounds.contains(newChildNode->logicalBounds())) {
-                        return UpdateLinkedGroupsError::UpdateExceedsWorldBounds;
+                ));
+            });
+
+            if (transformFailed) {
+                return UpdateLinkedGroupsError::TransformFailed;
+            }
+
+            // Move into map for easier lookup
+            auto resultsMap = std::unordered_map<const Model::Node*, NodeContentType>{};
+            for (auto& [origNode, transformedContents] : origNodeAndTransformedContents) {
+                resultsMap[origNode] = std::move(transformedContents);
+            }
+            origNodeAndTransformedContents.clear();
+
+            // Do a recursive traversal of the input node tree again,
+            // creating a matching tree structure, and move in the contents
+            // we've transformed above.
+
+            bool worldBoundsExceeded = false;
+            std::function<std::unique_ptr<Model::Node>(const Model::Node*)> cloneAndTransformRecursive = 
```

Comment:
Why not auto here? The signature is right below it, so I don't think this helps to make it more readable.

Reply:
AFAIK, it's necessary to declare it this way because it's a recursive lambda.

Reply:
But yeah, I agree, I wasn't doing it for readability.

Reply:
Ah yeah, I didn't make the connection at all. Looking at the code now, I wonder if it isn't worth just extracting it into a separate function though. What do you think of e99989539?

Reply:
Yeah, that's fine